ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/AggregateImpressions.java

```diff
@@ -0,0 +1,109 @@
+      ReportingUrlUtil.PARAM_FORM_FACTOR, ReportingUrlUtil.PARAM_OS_FAMILY,
+      ReportingUrlUtil.PARAM_ID);
+
+  private final ValueProvider<Integer> getAggregationWindowSize;
+
+  public AggregateImpressions(ValueProvider<Integer> getAggregationWindowSize) {
+    this.getAggregationWindowSize = getAggregationWindowSize;
+  }
+
+  public static AggregateImpressions of(ValueProvider<Integer> getAggregationWindowSize) {
+    return new AggregateImpressions(getAggregationWindowSize);
+  }
+
+  @Override
+  public PCollection<PubsubMessage> expand(PCollection<PubsubMessage> messages) {
+    return messages.apply(ParDo.of(new AddAggregationKey())) //
+        .apply(
+            Window.into(FixedWindows.of(Duration.standardMinutes(getAggregationWindowSize.get())))) //
+        .apply(GroupByKey.create()) //
+        .apply(ParDo.of(new CountImpressionsPerKey())) //
+        .apply(Window.into(new GlobalWindows()));
+  }
+
+  private class AddAggregationKey extends DoFn<PubsubMessage, KV<String, PubsubMessage>> {
+
+    @ProcessElement
+    public void processElement(@Element PubsubMessage message,
+        OutputReceiver<KV<String, PubsubMessage>> out) {
+      message = PubsubConstraints.ensureNonNull(message);
+
+      String reportingUrl = message.getAttribute(Attribute.REPORTING_URL);
+
+      ReportingUrlUtil urlParser = new ReportingUrlUtil(reportingUrl);
+
+      // Rebuild url filtering out unneeded parameters
+      ReportingUrlUtil aggregationUrl = new ReportingUrlUtil(urlParser.getBaseUrl());
+      for (String name : aggregationFields) {
+        aggregationUrl.addQueryParam(name, urlParser.getQueryParam(name));
+      }
+
+      out.output(KV.of(aggregationUrl.getReportingUrl().toString(), message));
+    }
+  }
+
+  private class CountImpressionsPerKey
+      extends DoFn<KV<String, Iterable<PubsubMessage>>, PubsubMessage> {
+
+    @ProcessElement
+    public void processElement(@Element KV<String, Iterable<PubsubMessage>> input,
+        OutputReceiver<PubsubMessage> out, IntervalWindow window) {
+      ReportingUrlUtil urlParser = new ReportingUrlUtil(input.getKey());
+
+      int impressionCount = Iterables.size(input.getValue());
+      long windowStart = window.start().getMillis();
+      long windowEnd = window.end().getMillis();
+
+      urlParser.addQueryParam(ReportingUrlUtil.PARAM_IMPRESSIONS,
+          Integer.toString(impressionCount));
+      urlParser.addQueryParam(ReportingUrlUtil.PARAM_TIMESTAMP_BEGIN,
+          Long.toString(windowStart / 1000));
+      urlParser.addQueryParam(ReportingUrlUtil.PARAM_TIMESTAMP_END,
+          Long.toString(windowEnd / 1000));
+
+      Map<String, String> attributeMap = ImmutableMap.of(Attribute.REPORTING_URL,
+          urlParser.toString());
+
+      ObjectNode json = Json.createObjectNode(); // TODO: for debug output
+      json.put(Attribute.REPORTING_URL, urlParser.toString());
+      json.put(Attribute.SUBMISSION_TIMESTAMP, Time.epochNanosToTimestamp(windowEnd / 1000));
```

Comment:
I need to think more about this too, but I'm not sure submission_timestamp gives us any real value here, and it may be better to let this job use processing time for windowing. That way, we'll never have to worry about late records, etc.

Reply:
The submission timestamp here is just for the bigquery output for debugging and the timestamp is just to have something easy to look at when looking at the table.  I think the windowing is using processing time but I was confused because I saw that late data was being dropped.

This is what shows up in the counters:
![Screen Shot 2021-05-12 at 5 28 42 PM](https://user-images.githubusercontent.com/12437227/118046660-9fcd1680-b347-11eb-9f63-637a02ff2be0.png)
This is from yesterday and I haven't been able to recreate it today.